In [2]:
print("Hello Jupyter!")

Hello Jupyter!


In [4]:
import time 
import math 

print("Exact value")

print("In order to get the exact value, in arraySum[] we are initiliazing an empty with a given size. Since we have 6 values we are trying to obtain, arraySum[] is initilizing all the 6 spaces with 0's.As the for loop goes through all the indexes 3 - 9, it performs an operation to get the sum. It is multiplying 10^i value with square root of 19. These actual sums will be used to compare the relative and absolute error.") 

exactSum = [0] * 6
for i in range(3, 9):          
    N = pow (10, i)
    exactSum[i-6] = N * math.sqrt(19.0)
    print('When N = 10^{} Exact Sum = {}'.format(i, exactSum[i-6]))
    
#Naive summation 
print('\n Naive Summation:')
def naiveSum(power):
    N = pow(10, power)
    s = 0
    start_time = time.time()
    for i in range(0, N):
        s += math.sqrt(19.0)
    end_time = time.time()
    elapsed_time = end_time - start_time
    index = power - 6 
    relativeError = abs((s-exactSum[index]) / exactSum[index])
    absError = abs(s - exactSum[index])
    print('N = 10 ^ {}\n\t s = {}  Elapsed time: {} secs  Absolute error: {}  Relative error: {}'.format(index +6, s, elapsed_time, absError, relativeError))
for  i in range(3, 9):
    naiveSum(i)
  

#Compensated Summation 
print("\nCompensated Summation")

def compensatedSum(power):
      N = pow(10, power)
      s = 0
      e = 0
      start_time = time.time()
      for i in range (0, N):
          temp = s
          y = math.sqrt(19.0) + e
          s = temp + y
          e = (temp -s) + y
          s += e
      end_time = time.time()
      elapsed_time = end_time - start_time 
      index = power - 6
      relativeError = abs((s-exactSum[index]) / exactSum[index])
      absError = abs(s - exactSum[index])
      print('N = 10 ^ {}\n\t s = {}  Elapsed time: {} secs  Absolute error: {}  Relative error: {}'.format(index +6, s, elapsed_time, absError, relativeError))
for  i in range(3, 9):
    compensatedSum(i)
    
      
    

Exact value
In order to get the exact value, in arraySum[] we are initiliazing an empty with a given size. Since we have 6 values we are trying to obtain, arraySum[] is initilizing all the 6 spaces with 0's.As the for loop goes through all the indexes 3 - 9, it performs an operation to get the sum. It is multiplying 10^i value with square root of 19. These actual sums will be used to compare the relative and absolute error.
When N = 10^3 Exact Sum = 4358.898943540674
When N = 10^4 Exact Sum = 43588.98943540674
When N = 10^5 Exact Sum = 435889.8943540674
When N = 10^6 Exact Sum = 4358898.943540674
When N = 10^7 Exact Sum = 43588989.43540674
When N = 10^8 Exact Sum = 435889894.3540674

 Naive Summation:
N = 10 ^ 3
	 s = 4358.898943540652  Elapsed time: 0.00023293495178222656 secs  Absolute error: 2.1827872842550278e-11  Relative error: 5.007657467006977e-15
N = 10 ^ 4
	 s = 43588.98943540103  Elapsed time: 0.002315998077392578 secs  Absolute error: 5.704350769519806e-09  Relative error: 

# Naive Summation
|N|Summation|Exact Value|Elapsed Time(s)|Absolute Error|Relative Error|
|-|---------|-----------|---------------|--------------|--------------|
|10^3|4358.898943540652|4358.898943540674|0.00012230873107910156|2.1827872842550278e-11|5.007657467006977e-15|
|10^4|43588.98943540103|43588.98943540674|.0014710426330566406|5.704350769519806e-09|1.30866781804449e-13|
|10^5|435889.8943536333|435889.8943540674|0.022168874740600586|4.341127350926399e-07|9.959229170383476e-13|
|10^6|4358898.943627511|4358898.943540674|0.14548110961914062|8.683744817972183e-05|1.9921876901597763e-11|
|10^7|43588989.44469575|43588989.43540674|.2897469997406006|0.009289011359214783|2.1310453579062438e-10|
|10^8|435889893.87809396|435889894.3540674|11.983985185623169|0.4759734272956848|1.09195793125926e-09|

# Compensation Summation
|N|Summation|Exact Value|Elapsed Time(s)|Absolute Error|Relative Error|
|-|---------|-----------|---------------|--------------|--------------|
|10^3|4358.898943540674|4358.898943540674|0.0001761913299560547|0.0|0.0|
|10^4|43588.98943540674|43588.98943540674|0.0017557144165039062 |0.0|0.0|
|10^5|435889.8943540674|435889.8943540674|0.019221067428588867|0.0|0.0|
|10^6|4358898.943540674|4358898.943540674|.1832139492034912 |0.0|0.0|
|10^7|43588989.43540674|43588989.43540674|1.9347419738769531|0.0|0.0|
|10^8|435889894.3540674|435889894.3540674|20.13253402709961|0.0|0.0|

The code above is calculating summation using two method, naive summation and compensated summation. The table compsres the values for each sumamtion.In summary, the naive summation values were far off than the exact value. As a result the absolute and relative errors values were bigger in number. On the other hand, the compensated summation methods summation was exactly the same as the exact values causing the absulute and relative errors to be zero all thoughout the 6 computations, 10^3, 10^4, 10^5, 10^6, 10^7, 10^7, 10^8. Since compensation summation is a recursive summation it was designed to diminish rounding errors. The correction term e is added to s before the next sum is added. As we discussed in class, the rounded floating-point arithmetic in base 2, e aka the correction term represents the error exactly. Additionally, the elapsed time for compensation summation took longer than naive summation because compensation summation is a recursive summation with a correction term, so it takes longer but with more accurate results.


Note: Running time may differ everytime the program is ran. The markdown tables presented here are data based on my past run. Though the elapsed time may differ by a few nanoseconds, the absolute and relative error are the same in all the outputs.